ssh/tailssh/incubator.go

```diff
@@ -104,7 +104,22 @@ func (ss *sshSession) newIncubatorCommand(logf logger.Logf) (cmd *exec.Cmd, err
 		loginShell := ss.conn.localUser.LoginShell()
 		args := shellArgs(isShell, ss.RawCommand())
 		logf("directly running %s %q", loginShell, args)
-		return exec.CommandContext(ss.ctx, loginShell, args...), nil
+		cmd = exec.CommandContext(ss.ctx, loginShell, args...)
+
+		// While running directly instead of using `tailscaled be-child`,
+		// do what sshd does by running inside the home directory if it exists.
+		homeDir := ss.conn.localUser.HomeDir
+		if _, err := os.Stat(homeDir); err == nil {
+			cmd.Dir = homeDir
+		} else if os.IsNotExist(err) {
+			// If the home directory doesn't exist, we can't chdir to it.
+			// Instead, we'll chdir to the root directory.
+			cmd.Dir = "/"
+		} else {
+			return nil, err
+		}
```

Comment:
@bradfitz was concerned about this kind of existence test in https://github.com/tailscale/tailscale/pull/13171#discussion_r1725168173. Now that I think about it, there must be a better way to determine if the home directory is useful.

Reply:
I’ve introduced `tryExecInDir` which should basically do what we want:
https://github.com/tailscale/tailscale/blob/a79da6ca31880428f0be7c7139bdfc4b8bd46e73/ssh/tailssh/incubator.go#L76-L95

We call `tryExecInDir` with the home directory, which will try to run a common command like `true`. This will fail if the home directory isn’t usable. Or it might fail if `true` isn’t available, because the system has been stripped down.

For reference, here is the full [diff](https://github.com/tailscale/tailscale/compare/e3dbf850c47912d752a51bdcec84c6acf29e1ade..a79da6ca31880428f0be7c7139bdfc4b8bd46e73) that addresses this comment.

The much simpler alternative is to _not_ fallback to the root directory when running directly. I wish I knew why the root fallback exists for this case.

Reply:
#13120 says that this is for root-squashed homedirs, which makes sense. But what does sshd do when the user’s homedir doesn’t exist or is not executable?

Reply:
SSHD falls back to root, so we should behave in the same way. I made `tryExecInDir` more paranoid by hardcoding absolute paths to these utilities. If this breaks in the future because /bin/true is missing, we could implement some kind of path-searching via os.Lookup.

Reply:
I’ve minimized the chance that someone hijacks the command by injecting something earlier in $PATH by hardcoding the test executables: https://github.com/tailscale/tailscale/pull/15351/commits/fef5c4872164d1ce3248d51272a41fd8c3aeb017#diff-5f1a2a45f6e2d0d5d45e93930bf6a3705761f390cc95e26c99aecce59c8f9236R87-R92

Reply:
After some more though, I think that a barebones system without `/bin/true` probably also doesn’t have home directories.